firedrake/slate/slate.py

```diff
@@ -444,6 +468,84 @@ def _key(self):
         return (type(self), self._function)
 
 
+class BlockAssembledVector(AssembledVector):
+    """This class is a symbolic representation of an assembled
+    vector of data contained in a set of :class:`firedrake.Function`s
+    defined on pieces of a split mixed function space.
+
+    :arg functions: A tuple of firedrake functions.
+    """
+    def __new__(cls, function, split_functions, indices):
+        if isinstance(split_functions, tuple) \
+           and all(isinstance(f, Coefficient) for f in split_functions):
+            self = TensorBase.__new__(cls)
+            self._function = split_functions
+            self._indices = indices
+            self._original_function = function
+            return self
+        else:
+            raise TypeError("Expecting a tuple of Coefficients or AssembledVectors (not a %r)" %
+                            type(split_functions))
+
+    @property
+    def coeff_map(self):
+        """A map from local coefficient numbers to the split global coefficient numbers.
+        The split coefficients are defined on the pieces of the originally mixed function spaces.
+        For BlockAssembledVectors the amount of split coefficients of an original coefficient
+        is not necessarily equal to the amount of split coefficients associated with the expression.
+        In this case the amount of split coefficients is dictated by how many pieces of the original
+        tensor the block is build on, this is defined by the indices.
+        """
+        sm = super().coeff_map()
+        pos = super().coefficients().index(self._original_function)
+        sm[pos] = self._indices
```

Comment:
I _think_ this won't work because `sm` has type `tuple` and is therefore immutable?

Reply:
Yes you are right. This functionality was actually untested, because the coeff map of the `BlockAssembledVector` is only called, when it's the outermost node. I added a test for that.
After looking at this again today, I also think we don't need this function at all. Picking the correct coefficients is already happening in https://github.com/firedrakeproject/firedrake/blob/520938d0b039854fbbf035d5081f1c35a93827c1/firedrake/slate/slate.py#L677